### DISCUSSION TITLE
Yaw optimization with disabled turbines

### DISCUSSION THREAD

**Discussion by MiguelMarante** [NatLabRockies/floris#1016] [General]
Hello everyone,

I want to implement yaw optimization for a time-series, taking into account that different turbines are not operational at different time records.

Below is an example for a time-series with two records: the mid-stream turbine is disabled in the first record, and the upstream turbine is disabled in the second. The results show that this information is ignored in the optimization procedure.

Is there a workaround to ensure these turbines are completely excluded from the optimization?

**Note:** I am using Floris v4.1 in this example.

Best regards,
Miguel Marante

CODE:
```python
import numpy as np

from floris import FlorisModel
from floris.optimization.yaw_optimization.yaw_optimizer_sr import YawOptimizationSR

# Create floris object
yaml_file = 'inputs/emgauss.yaml'
fmodel = FlorisModel(yaml_file)

# Set wind conditions
fmodel.set(wind_directions=[270.]*2, wind_speeds=[8.]*2, turbulence_intensities=[.06]*2)

# Disable turbines
fmodel.set_operation_model('mixed')
fmodel.set(disable_turbines=np.array([[False, True, False], [True, False, False]]))

# Run optimization and print results
yaw_opt = YawOptimizationSR(fmodel)
df_opt = yaw_opt.optimize()
print(df_opt)
```

![image](https://github.com/user-attachments/assets/cf4a2802-da6e-45d1-983a-faa0ecab5f6e)

**Comment by misi9170** [comment]
Hi @MiguelMarante ,

Thank you for raising this and sharing your example code. This is a bug, to do with the `power_setpoints` (which are set to near zero when turbines are disabled) not being correctly applied on a copy of the `FlorisModel` created within the yaw optimization routine.

I'll try to work on a bugfix in the coming days. In the meantime, you may be able to create a workaround by using a combination of the `turbine_weights` and `minimum_yaw_angle` and `maximum_yaw_angle` keyword arguments; however, you'll then have to run a single yaw optimization per time step, which will be clunky and slow.

Again, thanks for bringing this to our attention. I hope we'll be able to provide a suitable fix reasonably quickly.

Misha

**Reply by MiguelMarante** [reply]
Hi @misi9170,

Thank you for the quick response and for looking into this. Do you think this fix will come in the next release? And if so, do you have a prevision for when that will be?

Best regards,
Miguel Pereira

**Reply by misi9170** [reply]
Hi Miguel,

I'm working on a fix now and hoping to have something working by the end of the day. That will be a new pull request, and will hopefully be merged into the develop branch in the coming days. Then, depending on the status of a couple of other issues/pull requests, we are hoping to get a new patch release of FLORIS (4.2.1) done by the end of next week or possibly sooner.

You'll hopefully be able to try things out anywhere in that process once I open the PR---either on the PR branch, once it's in the develop branch, or once it makes it into main and v4.2.1 is released.

I'll update this comment with progress on that.

Cheers,
Misha

___
UPDATE: I have now created a fix for this in #1031 (previously #1027) and merged it into the develop branch.

**Reply by misi9170** [reply]
The bug fix in #1031 is now on the FLORIS main branch and released as part of [v4.2.1](https://github.com/NREL/floris/releases/tag/v4.2.1)

**Reply by MiguelMarante** [reply]
Hey @misi9170,

I updated the Floris version to 4.2.1 (I was previously using Floris 4.1) and I started to get these warnings from the Serial Refinement optimization. Am I doing something wrong here?

```python
import numpy as np

from floris import FlorisModel, WindRose
from floris.optimization.yaw_optimization.yaw_optimizer_sr import YawOptimizationSR

wind_directions = np.array([270.,272.])
wind_speeds = np.array([8.,9.])

wind_rose = WindRose(wind_directions=wind_directions, wind_speeds=wind_speeds, ti_table=0.06)

fmodel = FlorisModel("inputs_floris/emgauss.yaml")
fmodel.set(wind_data=wind_rose)

yaw_opt = YawOptimizationSR(fmodel)
df_opt = yaw_opt.optimize()
```

![image](https://github.com/user-attachments/assets/044d886c-4a2c-4315-ac17-ce05db97d04e)

**Reply by misi9170** [reply]
Hi @MiguelMarante ,

Thanks for bringing this up, and sorry you are continuing to have issues here. Thanks also for providing a working script for reproducing the issue.

I've looked into this quickly, and I believe you can ignore the warning. It is being produced now as an unintended consequence of the way that I addressed the earlier issue in #1031 , but it is disconcerting and I will to create a second small fix to prevent this from happening. Thanks for bringing it to my attention.

Misha

**Reply by misi9170** [reply]
Hi @MiguelMarante ,

We just released [v4.2.2](https://github.com/NREL/floris/releases/tag/v4.2.2), which should address those unnecessary warnings. If you update your FLORIS installation, you should see those go away. See #1045 for details.

Misha
